File: lc/coreprep/implementation/practice_iterations/patterns_iterations_1/search_in_sorted_infinite_array.cpp
#include <algorithm>
#include <iostream>
#include <limits>
#include <string>
#include <vector>

using namespace std;

struct ArrayReader {
  const vector<int> a;

  ArrayReader(const vector<int> a) : a(std::move(a)) {}

  int get(int idx) const {
    if (idx >= a.size()) {
      return numeric_limits<int>::max();
    }
    return a[idx];
  }
};

// O(logn) / O(1)
int binary_search(const ArrayReader &a, int k) {
  int low = 0, high = 1;

  while (a.get(high) < k) {
    int new_low = high + 1;
    high += (high - low + 1) * 2;
    low = new_low;
  }

  while (low <= high) {
    int mid = low + (high - low) / 2;

    if (a.get(mid) < k) {
      low = mid + 1;
    } else if (a.get(mid) > k) {
      high = mid - 1;
    } else {
      return mid;
    }
  }

  return -1;
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n, k;
  cin >> tt;

  while (tt--) {
    cin >> n >> k;

    vector<int> a(n);
    for (int i = 0; i < n; i++) {
      cin >> a[i];
    }

    ArrayReader arr(a);
    cout << binary_search(arr, k) << "\n";
  }

  return 0;
}

File: lc/coreprep/implementation/practice_iterations/patterns_iterations_2/SearchInSortedInfiniteArray.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class SearchInSortedInfiniteArray {
  static class ArrayReader {
    int[] a;

    ArrayReader(int[] a) { this.a = a; }

    int get(int idx) {
      if (idx >= a.length) {
        return Integer.MAX_VALUE;
      }
      return a[idx];
    }
  }

  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();
        int k = in.nextInt();

        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
          a[i] = in.nextInt();
        }

        ArrayReader arr = new ArrayReader(a);
        System.out.println(binarySearch(arr, k));
      }
    }
  }

  // O(logn) / O(1)
  public static int binarySearch(ArrayReader arr, int k) {
    int low = 0, high = 1;
    while (arr.get(high) < k) {
      low = high;
      high <<= 1;
    }

    while (low <= high) {
      int mid = low + (high - low) / 2;

      if (arr.get(mid) < k) {
        low = mid + 1;
      } else if (arr.get(mid) > k) {
        high = mid - 1;
      } else {
        return mid;
      }
    }

    return -1;
  }
}